### PR TITLE
feat: cli improvement

### LINKED ISSUE #47: Make the CLI interface smarter
Currently, we need to add exact paths in the cli. See this `python3 starfyre --dev=True --path="/Users/sanskar/repos/starfyre/test-application/" && python -m starfyre --build=True --path="/Users/sanskar/repos/starfyre/test-application/"`

Make this cli more human friendly. 

- remove the need to pass =True
- make this sequential invocation into a single command
- make the `--path` smarter

From our readme
```
Run the script `./build.sh`. This command will run the build process in starfyre against the test application in `test-application` directory.
  - The `build.sh` file is a simple script that runs two commands sequentially.
    - `python -m starfyre --dev=True --path="test-application/"`
    - `python -m starfyre --build=True --path="test-application/"`
        - The `path` variable here is the path to our application.
        - The `dev` flag here is used to start the compilation and create the `build` directory. 
        - The `build` directory is basically a python package that contains all the compiled files. We use the `--build` flag to run that package.
```

These both commands should be a single command by default.

### PR DESCRIPTION
**Pull request by DeleMike**
Closes #47 
I have done some research and added `--dev/--no-dev` and `--build/--no-build` instead of the former commands. This new option style allows the build file to now has `python3 starfyre --build --path="test-application"`. It makes the command leaner.

If the user runs the build file,  it will successfully build and run the required files to create our application.

Things Observed:
If I change the build file from `python3 starfyre --build --path="test-application"` to `python3 starfyre --build --path="my-first-app"`. It will report the following error:

```
 File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/akin/software_eng/starfyre/starfyre/__main__.py", line 74, in <module>
    main()
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/akin/software_eng/starfyre/starfyre/__main__.py", line 61, in main
    compile(path)
  File "/home/akin/software_eng/starfyre/starfyre/compiler.py", line 160, in compile
    build_dir.mkdir(exist_ok=True)
  File "/usr/local/lib/python3.10/pathlib.py", line 1175, in mkdir
    self._accessor.mkdir(self, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/akin/software_eng/starfyre/my-first-app/build'
```
Is this an expected behaviour?

Is the `--path="test_application"` in the command contained in the starfyre build file constant or `path` value can change?

### CODE REVIEW

File: starfyre/__main__.py [hunk 1]
```diff
@@ -42,16 +42,26 @@ def create_main_file(path):
 
 @click.command()
 @click.option("--path", default=".", help="Path to the project")
-@click.option("--dev", default=False, help="Start the compilation and generate the build package.")
-@click.option("--build", default=False, help="Start the build package")
+@click.option("--dev/--no-dev", default=False, help="Start the compilation and generate the build package.")
+@click.option("--build/--no-build", default=False, help="Start the build package")
```
Comment (sansyrox) [inline starfyre/__main__.py:46]: What is the purpose of having a `--no-dev` or a `--no-build` flag here?? It makes the interface a bit more confusing imo.

What we are looking for is the `is_flag` argument. You can read more about it here(https://pymbook.readthedocs.io/en/latest/click.html)

Comment (sansyrox) [inline starfyre/__main__.py:46]: What is the purpose of having a --no-dev or a --no-build flag here?? It makes the interface a bit more confusing imo.

What we are looking for is the is_flag argument. You can read more about it here(https://pymbook.readthedocs.io/en/latest/click.html)

File: starfyre/__main__.py [hunk 2]
```diff
@@ -42,16 +42,26 @@ def create_main_file(path):
...
 def main(path, dev, build):
+    """
+    Command-line interface to compile and build a project.
+
+    Args:
+        path (str): Path to the project directory.
+        dev (bool): Whether to start the compilation and generate the build package.
+        build (bool): Whether to start the build package.
```
Comment (sansyrox) [inline starfyre/__main__.py:54]: Can we have a `\n` at the end of lines here? As this docstring is visible when you type the command `python -m starfyre --help`.

So, e.g. 
```
path (str): Path to the project directory.\n
dev (bool): Whether to start the compilation and generate the build package.\n
```

File: build.sh [hunk 3]
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
 
-python3 starfyre --dev=True --path="test-application/" && python -m starfyre --build=True --path="test-application/"
+python3 starfyre --build --path="test-application"
```
Comment (sansyrox) [inline build.sh:3]: One suggestion - we should remove the `dev` flag from the build script once we remove it from the `main()` function.

**Review by sansyrox — COMMENTED** [review]
Hey @DeleMike 👋 

Great work! I love the speed of PR 🔥 

I have a few comments on the PR.  Do let me know if you require any more information from me. 😄

**Comment by sansyrox** [comment]
Also, answering your question here.

> If I change the build file from python3 starfyre --build --path="test-application" to python3 starfyre --build --path="my-first-app". It will report the following error:

It is because it will either expect a relative path or an absolute path. So, it is unable to find the folder as you do not have any folder named `my-first-app` in your starfyre directory.

You can decipher it from the error message here

> FileNotFoundError: [Errno 2] No such file or directory: '/home/akin/software_eng/starfyre/my-first-app/build'

So, it is not a fixed value but this is the expected error.

File: starfyre/__main__.py (hunk 1 above)
Comment (DeleMike) [reply starfyre/__main__.py:46]: Thanks for this! Updated.

**Comment by DeleMike** [comment]
Hi @sansyrox, I have updated the implementation - I made some changes with your suggestions, thanks! The issue now is this command `python3 starfyre --path="my-first-app"` works fine but I think it doesn't generate all the files needed for a successful run, right?

I compared it against the `test_application` and it is missing some files. How do these files get generated?

So because `python3 starfyre --path="my-first-app"` doesn't generate all the files needed, this command `python3 starfyre --build --path="my-first-app"` fails with this error: 

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/akin/software_eng/starfyre/starfyre/__main__.py", line 75, in <module>
    main()
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/akin/software_eng/starfyre/.venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/akin/software_eng/starfyre/starfyre/__main__.py", line 66, in main
    subprocess.run(
  File "/usr/local/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/local/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/local/lib/python3.10/subprocess.py", line 1847, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/akin/software_eng/starfyre/my-first-app/build/__init__.py'

```
 I tried printing the absolute path and it is pointing to the right directory `Absolute Path = /home/akin/software_eng/starfyre/my-first-app/build/__init__.py`
 
 Am I missing a step during compilation?

**Comment by DeleMike** [comment]
I mistakenly deleted the whole project while testing if the compilation was working but I was able to get the files back later. 
I have cut out the section of code I was working with in the `__main__.py` file


```
@click.command()
@click.option("--path", default=".", help="Path to the project")
@click.option("--build", is_flag=True, help="Compile and build package")
def main(path, build):
    """
    Command-line interface to compile and build a project.
    Args:
        path (str): Path to the project directory.\n
        build (bool): Whether to start the build package.
    """
    # compile and build project
    temp_path = path
    path_ = temp_path + "/__init__.py"
    # get absolute path
    path = os.path.abspath(path_)
    compile(path)
    create_main_file(os.path.dirname(path))
    if build:
    # start/run project
        path_ = temp_path + "/build/__init__.py"
        path = os.path.abspath(path_)
        subprocess.run(
            [sys.executable, "-m", "build"],
            cwd=path,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )


if __name__ == "__main__":
    main()
```

**Comment by DeleMike** [comment]
Okay, so an update. I saw that the `compile` function expected an **absolute** path and I was passing a **relative** one - so the issue is not fixed. If you run `python3 starfyre --build --path="my-first-app"` or `python3 starfyre --path="my-first-app"`, they will run successfully. 

The only question I have now is the build method only produced this file structure:
```
    my-first-app/
        build/
            __main__.py
        dist/
            store.js
```

are there not meant to be more files?

File: .gitignore
Comment (sansyrox) [inline .gitignore:1]: It is usually not a good practice to add non generic files in the `.gitignore`. Can you please remove this?

File: build.sh [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
 
-python3 starfyre --dev=True --path="test-application/" && python -m starfyre --build=True --path="test-application/"
+python3 starfyre --build --path="my-first-app"
```
Comment (sansyrox) [inline build.sh:3]: We do not have a folder called `my-first-app` in our project. It should still be`test-application` for us.

File: starfyre/__main__.py [hunk 5]
```diff
@@ -1,64 +1,73 @@
...
+    Args:
+        path (str): Path to the project directory.
+        build (bool): Whether to start the build package.
+    """
+    # Convert path to absolute path
+    path = os.path.abspath(path)
+
+    # Compile and build project
+    compile(os.path.join(path, "__init__.py"))
+    create_main_file(path)
```
Comment (sansyrox) [inline starfyre/__main__.py:59]: I don't think that we should be compiling the project if the `--build` flag has not been supplied.

File: starfyre/__main__.py [hunk 6]
```diff
@@ -1,64 +1,73 @@
...
+def main(path, build):
+    """
+    Command-line interface to compile and build a project.
+
+    Args:
+        path (str): Path to the project directory.
+        build (bool): Whether to start the build package.
+    """
+    # Convert path to absolute path
+    path = os.path.abspath(path)
```
Comment (sansyrox) [inline starfyre/__main__.py:55]: Also, can we add a test to check if the `--path` flag has been supplied? This will result in an error otherwise.

**Review by sansyrox — COMMENTED** [review]
Hey @DeleMike 👋 

Thank you for addressing the changes. However, I have a few more comments on this PR. Do let me know if you require any more clarification 😄

**Comment by sansyrox** [comment]
> Okay, so an update. I saw that the compile function expected an absolute path and I was passing a relative one - so the issue is not fixed. If you run python3 starfyre --build --path="my-first-app" or python3 starfyre --path="my-first-app", they will run successfully.
The only question I have now is the build method only produced this file structure:
are there not meant to be more files?

@DeleMike , there should be more files, given that we have merged the dev and the build flags. The `.fyre` files are compiled to `.py` files.

File: .gitignore
Comment (DeleMike) [reply .gitignore:1]: removed.

File: starfyre/__main__.py (hunk 6 above)
Comment (DeleMike) [reply starfyre/__main__.py]: additional functionality added.

**Comment by DeleMike** [comment]
> @DeleMike , there should be more files, given that we have merged the dev and the build flags. The `.fyre` files are compiled to `.py` files.

@sansyrox  Can you please check this: 
```python
 if build:
        # Start/run project
        build_dir = os.path.join(path, "build")
        subprocess.run(
            [sys.executable, "-m", "build"],
            cwd=build_dir,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    else:
        # Compile and build project
        compile(os.path.join(path, "__init__.py"))
        create_main_file(path)
``` 
Does the above actually represent that we have merged the two flags? You suggested that I do not compile if `--build` is not passed. If I shouldn't compile what tips do you have for me to merge both commands?

File: starfyre/__main__.py [hunk 7]
```diff
@@ -41,23 +41,39 @@ def create_main_file(path):
...
+
+    Args:
+
+        path (str): Path to the project directory.\n
+        build (bool): Whether to start the build package.
+    """
+    if not path:
+        click.echo("Error: Please provide a valid path using the --path flag.\nUse --help for more details")
+        return
+
```
Comment (sansyrox) [inline starfyre/__main__.py:58]: Nice! ✨

File: starfyre/__main__.py [hunk 8]
```diff
@@ -41,23 +41,39 @@ def create_main_file(path):
...
+
+    # Convert path to absolute path
+    path = os.path.abspath(path)
```
Comment (sansyrox) [inline starfyre/__main__.py:61]: One suggestion - we will be writing cleaner code if start using Path from pathlib(https://docs.python.org/3/library/pathlib.html)

File: starfyre/__main__.py [hunk 9]
```diff
@@ -41,23 +41,39 @@ def create_main_file(path):
...
+        # Start/run project
+        build_dir = os.path.join(path, "build")
         subprocess.run(
             [sys.executable, "-m", "build"],
-            cwd=path,
+            cwd=build_dir,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,
         )
+    else:
```
Comment (sansyrox) [inline starfyre/__main__.py:72]: Let us remove this `else` here. We shouldn't only compile if the build command is not passed. At least for now

**Review by sansyrox — COMMENTED** [review]
Hi @DeleMike 👋 

Great work. I have left a few suggestions. 

Also, I will be pushing a commit to fix the large diffs soon 😄

**Comment by DeleMike** [comment]
Hello @sansyrox , I have made some updates.

**Comment by sansyrox** [comment]
Thanks @DeleMike 😄 

I have made a few changes finally. And this should be good to merge. Great work! 🔥

**Comment by sansyrox** [comment]
Actually @DeleMike , one update. The build script has stopped working.

If you delete the build and the dist folder. And then re start the `./build` script. There is no index.html in the `dist` folder. There must be a bug somewhere. Can you please have a look ?

**Comment by DeleMike** [comment]
Thanks. I'm looking into it now...

File: starfyre/__main__.py [hunk 10]
```diff
@@ -41,20 +41,38 @@ def create_main_file(path):
...
+        build_dir = absolute_path / "build"
+        init_file_path = absolute_path / "__init__.py"
+        compile(init_file_path.resolve())
+        create_main_file(str(absolute_path))
+
+        # Start/run project
         subprocess.run(
             [sys.executable, "-m", "build"],
-            cwd=path,
+            cwd=build_dir.resolve(),
```
Comment (sansyrox) [inline starfyre/__main__.py:75]: @DeleMike , this is the issue here. This needs to be path and not `build_dir`

Comment (DeleMike) [reply starfyre/__main__.py:75]: Thanks, let me look into it more!

**Pull request merged**